entity-service-impl/src/main/java/org/hypertrace/entity/query/service/EntityQueryServiceImpl.java

```diff
@@ -290,19 +296,115 @@ private void doUpdate(RequestContext requestContext, EntityUpdateRequest request
-          LOG.warn(
-              "Failed to update entity {}, subDocPath {}, with new doc {}.",
-              key,
-              subDocPath,
-              jsonValue);
+        if (toUpdate.containsValue(key)) {
+          toUpdate.get(key).put(subDocPath, new JSONDocument(jsonValue));
+        } else {
+          Map<String, Document> subDocument = new HashMap<>();
+          subDocument.put(subDocPath, new JSONDocument(jsonValue));
+          toUpdate.put(key, subDocument);
         }
       }
+      try {
+        entitiesCollection.bulkUpdateSubDocs(toUpdate);
+      } catch (Exception e) {
+        LOG.warn(
+            "Failed to update entities, subDocPath {}, with new doc {}.", subDocPath, jsonValue);
+      }
+    }
+  }
+
+  @Override
+  public void bulkUpdate(
+      BulkEntityUpdateRequest request, StreamObserver<ResultSetChunk> responseObserver) {
+    // Validations
+    RequestContext requestContext = RequestContext.CURRENT.get();
+    Optional<String> tenantId = requestContext.getTenantId();
+    if (tenantId.isEmpty()) {
+      responseObserver.onError(new ServiceException("Tenant id is missing in the request."));
+      return;
+    }
+    if (StringUtils.isEmpty(request.getEntityType())) {
+      responseObserver.onError(new ServiceException("Entity type is missing in the request."));
+      return;
+    }
+    if (request.getEntitiesCount() == 0) {
+      responseObserver.onError(new ServiceException("Entity IDs are missing in the request."));
+    }
+    try {
+      bulkDoUpdate(requestContext, request);
+      // Finally return the selections
+      List<String> entityIdsList = new ArrayList<>();
+      for (String entityId : request.getEntities().keySet()) {
+        entityIdsList.add(entityId);
+      }
+      Query entitiesQuery = Query.newBuilder().addAllEntityId(entityIdsList).build();
+      List<String> docStoreSelections =
+          entityQueryConverter.convertSelectionsToDocStoreSelections(
+              requestContext, request.getSelectionList());
+      Iterator<Document> documentIterator =
+          entitiesCollection.search(
+              DocStoreConverter.transform(tenantId.get(), entitiesQuery, docStoreSelections));
+      List<Entity> entities = convertDocsToEntities(documentIterator);
+      responseObserver.onNext(
+          convertEntitiesToResultSetChunk(requestContext, entities, request.getSelectionList()));
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      responseObserver.onError(
+          new ServiceException("Error occurred while executing " + request, e));
+    }
+  }
+
+  private void bulkDoUpdate(RequestContext requestContext, BulkEntityUpdateRequest request) {
+    Map<Key, Map<String, Document>> toUpdate = new HashMap<>();
+    Map<String, Document> documentMap = new HashMap<>();
+    for (String entityId : request.getEntities().keySet()) {
+      documentMap.clear();
```

Comment:
We should in fact create the `documentMap` inside the for loop, instead of clearing it always. Once you move it inside the for loop, can you also please rename it to `entityDocumentMap`?

Reply:
I don't think you even need the document map. You can just use the output of the `documentMapMaker` and directly put it inside the `toUpdate` map